File: ios/GuideStreamTV/Services/TeamFavoritesService.swift
//
//  TeamFavoritesService.swift
//  GuideStreamTV
//
//  Persists favorited sports team uids to Supabase so the backend can
//  deliver starting-soon, going-live, and final-score push notifications
//  for games involving those teams. Scoped to the signed-in user or guest
//  device, mirroring the StreamsViewModel persistence pattern.
//

import Foundation
import Supabase

@MainActor
@Observable
final class TeamFavoritesService {
    static let shared = TeamFavoritesService()

    /// In-memory set of favorited team uids (ESPN's globally-unique uid).
    private(set) var favoritedUids: Set<String> = []

    /// Loaded rows from the team_favorites table, keyed by team_uid.
    /// Public so SportsView can read team_abbr/team_name/league/sport for chip rendering.
    private(set) var rows: [String: TeamFavoriteRow] = [:]

    /// Cached UserDefaults key for offline/guest survival.
    private let localCacheKey = "gs.teamFavorites.localCache.v1"

    private var currentUserId: UUID? {
        AuthViewModel.shared.currentUser?.id
    }

    private var deviceId: String {
        DeviceIdentity.shared.deviceId
    }

    private init() {
        favoritedUids = loadLocalCache()
    }

    // MARK: - Load

    /// Fetches the current user's or device's favorites from the server.
    func load() async {
        do {
            var query = SupabaseManager.shared.client
                .from("team_favorites")
                .select()
            if let uid = currentUserId?.uuidString {
                query = query.or("user_id.eq.\(uid),device_id.eq.\(deviceId)")
            } else {
                query = query.eq("device_id", value: deviceId)
            }
            let loaded: [TeamFavoriteRow] = try await query.execute().value
            var uids = Set<String>()
            var rowMap: [String: TeamFavoriteRow] = [:]
            for row in loaded {
                guard let teamUid = row.team_uid else { continue }
                uids.insert(teamUid)
                rowMap[teamUid] = row
            }
            self.favoritedUids = uids
            self.rows = rowMap
            saveLocalCache(uids)
        } catch {
            print("[TeamFavorites] load failed: \(error.localizedDescription)")
            // Keep showing local cache on failure.
            favoritedUids = loadLocalCache()
        }
    }

    // MARK: - Query

    /// Returns `true` when the given team uid is currently favorited.
    func isFavorite(_ uid: String?) -> Bool {
        guard let uid else { return false }
        return favoritedUids.contains(uid)
    }

    /// Returns the current set of favorited team uids.
    func favoriteUids() -> Set<String> {
        favoritedUids
    }

    // MARK: - Toggle

    /// Optimistically adds or removes a team favorite, then persists to Supabase.
    /// No-ops when `team.uid` is nil.
    func toggle(team: GameTeam, league: String?, sport: String?) async {
        guard let uid = team.uid else { return }

        let wasFavorited = favoritedUids.contains(uid)

        // Optimistic local update
        if wasFavorited {
            favoritedUids.remove(uid)
            rows.removeValue(forKey: uid)
        } else {
            favoritedUids.insert(uid)
        }
        saveLocalCache(favoritedUids)

        do {
            if wasFavorited {
                // Delete matching row(s) for this team_uid scoped to current user/device.
                var query = SupabaseManager.shared.client
                    .from("team_favorites")
                    .delete()
                    .eq("team_uid", value: uid)
                if let userId = currentUserId?.uuidString {
                    query = query.or("user_id.eq.\(userId),device_id.eq.\(deviceId)")
                } else {
                    query = query.eq("device_id", value: deviceId)
                }
                try await query.execute()
            } else {
                // Insert a row with the team details.
                var payload: [String: AnyJSON] = [
                    "device_id": .string(deviceId),
                    "team_uid": .string(uid)
                ]
                if let userId = currentUserId?.uuidString {
                    payload["user_id"] = .string(userId)
                }
                if let teamId = team.id { payload["team_id"] = .string(teamId) }
                if !team.abbreviation.isEmpty, team.abbreviation != "—" {
                    payload["team_abbr"] = .string(team.abbreviation)
                }
                payload["team_name"] = .string(team.shortName)
                if let league { payload["league"] = .string(league) }
                if let sport { payload["sport"] = .string(sport) }

                try await SupabaseManager.shared.client
                    .from("team_favorites")
                    .insert(payload)
                    .execute()
            }
            // When favoriting a team, ensure the user is set up for push notifications
            // via the existing PushTokenManager path so the backend can deliver
            // starting-soon, going-live, and final-score alerts.
            if !wasFavorited {
                await PushTokenManager.shared.resaveCachedToken()
            }
        } catch {
            let message = error.localizedDescription.lowercased()
            // Duplicate row → not really an error.
            if message.contains("duplicate") || message.contains("23505") {
                // Already saved — no-op.
            } else {
                print("[TeamFavorites] toggle persist failed: \(error.localizedDescription)")
                // Keep the optimistic local state — failures are logged but never
                // undo what the user sees, matching the StreamsViewModel pattern.
            }
        }
    }

    // MARK: - Local cache

    private func loadLocalCache() -> Set<String> {
        guard let data = UserDefaults.standard.data(forKey: localCacheKey) else { return [] }
        do {
            return try JSONDecoder().decode(Set<String>.self, from: data)
        } catch {
            return []
        }
    }

    private func saveLocalCache(_ uids: Set<String>) {
        do {
            let data = try JSONEncoder().encode(uids)
            UserDefaults.standard.set(data, forKey: localCacheKey)
        } catch {
            print("[TeamFavorites] local cache encode failed: \(error.localizedDescription)")
        }
    }
}

// MARK: - Decodable row for server reads

nonisolated struct TeamFavoriteRow: Decodable, Sendable {
    let team_uid: String?
    let team_id: String?
    let team_abbr: String?
    let team_name: String?
    let league: String?
    let sport: String?
    let user_id: String?
    let device_id: String?
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/repository/TeamFavoritesService.kt
package com.rork.guidestreamtvandroid.data.repository

import android.content.Context
import android.util.Log
import com.rork.guidestreamtvandroid.data.local.DeviceIdentity
import com.rork.guidestreamtvandroid.data.models.SportsGame
import com.rork.guidestreamtvandroid.data.remote.SupabaseManager
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Persists the user's favorited sports teams — mirrors iOS
 * TeamFavoritesService.swift. Keyed by ESPN team uid so a favorite survives
 * across game refreshes.
 *
 * Favorites are written through to `public.team_favorites` on Supabase as well
 * as to SharedPreferences. The server copy is what `sports_poll_and_notify`
 * reads to send starting-soon / going-live / final-score pushes — without it,
 * Android favorites are invisible to the backend and generate no notifications.
 * Scoping matches StreamsViewModel exactly: signed-in rows are keyed by
 * `user_id`, guest rows by `device_id` with `user_id IS NULL`, so two accounts
 * on one install never read or delete each other's favorites.
 *
 * Local writes are optimistic and are never rolled back on a network failure,
 * matching the iOS behaviour and the StreamsViewModel pattern.
 */
class TeamFavoritesService private constructor(context: Context) {

    @Serializable
    data class FavoriteRow(
        val teamUid: String,
        val teamAbbr: String? = null,
        val teamName: String? = null,
        val league: String? = null,
        val sport: String? = null,
        val teamId: String? = null,
    )

    /** Wire shape of a public.team_favorites row. */
    @Serializable
    private data class RemoteFavoriteRow(
        @SerialName("team_uid") val teamUid: String? = null,
        @SerialName("team_id") val teamId: String? = null,
        @SerialName("team_abbr") val teamAbbr: String? = null,
        @SerialName("team_name") val teamName: String? = null,
        val league: String? = null,
        val sport: String? = null,
    )

    private val prefs = context.getSharedPreferences("gs.teamFavorites", Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true }
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _rows = MutableStateFlow<Map<String, FavoriteRow>>(loadFromDisk())
    val rows: StateFlow<Map<String, FavoriteRow>> = _rows.asStateFlow()

    private val currentUserId: String? get() = AuthViewModel.get().currentUserId
    private val deviceId: String get() = DeviceIdentity.get().deviceId

    /** Favorited team uids in insertion order. */
    fun favoriteUids(): List<String> = _rows.value.keys.toList()

    fun isFavorite(uid: String?): Boolean = uid != null && _rows.value.containsKey(uid)

    /**
     * Loads the current owner's favorites from Supabase and replaces the local
     * cache with the result. Keeps the existing local cache on failure so the
     * chips never blank out because the network hiccuped.
     */
    suspend fun load() {
        try {
            val uid = currentUserId
            val remote = SupabaseManager.client.postgrest
                .from("team_favorites")
                .select {
                    filter {
                        if (uid != null) {
                            eq("user_id", uid)
                        } else {
                            eq("device_id", deviceId)
                            exact("user_id", null)
                        }
                    }
                }
                .decodeList<RemoteFavoriteRow>()

            val mapped = remote.mapNotNull { r ->
                val teamUid = r.teamUid ?: return@mapNotNull null
                teamUid to FavoriteRow(
                    teamUid = teamUid,
                    teamAbbr = r.teamAbbr,
                    teamName = r.teamName,
                    league = r.league,
                    sport = r.sport,
                    teamId = r.teamId,
                )
            }.toMap()

            _rows.value = mapped
            saveToDisk(mapped)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            Log.e(TAG, "load failed: ${e.message}", e)
            // Leave the local cache in place.
        }
    }

    /** Fire-and-forget wrapper for callers outside a coroutine scope. */
    fun refresh() {
        scope.launch { load() }
    }

    /** Toggles a team's favorite state. No-op if the team has no stable uid. */
    fun toggle(team: SportsGame.TeamSummary, league: String?, sport: String?) {
        val uid = team.uid ?: return
        val wasFavorited = _rows.value.containsKey(uid)
        val current = _rows.value.toMutableMap()

        if (wasFavorited) {
            current.remove(uid)
        } else {
            current[uid] = FavoriteRow(
                teamUid = uid,
                teamAbbr = team.abbreviation,
                teamName = team.shortName.ifEmpty { team.displayName.ifEmpty { team.name } },
                league = league,
                sport = sport,
            )
        }
        _rows.value = current
        saveToDisk(current)

        val inserted = current[uid]
        scope.launch {
            if (wasFavorited || inserted == null) deleteRemote(uid) else insertRemote(listOf(inserted))
        }
    }

    /**
     * Adds several favorites in one pass — used by the team picker, which can
     * commit an arbitrary number of teams at once. Already-favorited uids are
     * skipped, so this is safe to call with a full selection.
     */
    suspend fun addMany(entries: List<FavoriteRow>) {
        val current = _rows.value.toMutableMap()
        val added = entries.filter { !current.containsKey(it.teamUid) }
        if (added.isEmpty()) return

        added.forEach { current[it.teamUid] = it }
        _rows.value = current
        saveToDisk(current)

        insertRemote(added)
    }

    /** Removes several favorites in one pass. Missing uids are ignored. */
    suspend fun removeMany(uids: List<String>) {
        val current = _rows.value.toMutableMap()
        val removed = uids.filter { current.remove(it) != null }
        if (removed.isEmpty()) return

        _rows.value = current
        saveToDisk(current)

        removed.forEach { deleteRemote(it) }
    }

    private suspend fun insertRemote(entries: List<FavoriteRow>) {
        if (entries.isEmpty()) return
        val uid = currentUserId
        val payloads = entries.map { row ->
            buildJsonObject {
                put("device_id", deviceId)
                put("team_uid", row.teamUid)
                if (uid != null) put("user_id", uid)
                row.teamId?.let { put("team_id", it) }
                row.teamAbbr?.takeIf { it.isNotBlank() && it != "\u2014" }?.let { put("team_abbr", it) }
                row.teamName?.let { put("team_name", it) }
                row.league?.let { put("league", it) }
                row.sport?.let { put("sport", it) }
            }
        }
        try {
            SupabaseManager.client.postgrest.from("team_favorites").insert(payloads)
            // A favorite is only useful if the backend can reach this device.
            PushTokenManager.get().resaveCachedToken()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            val msg = e.message?.lowercase() ?: ""
            if (msg.contains("duplicate") || msg.contains("23505")) {
                // Already saved — no-op.
            } else {
                Log.e(TAG, "insert failed: ${e.message}", e)
                // Keep the optimistic local state, matching iOS.
            }
        }
    }

    private suspend fun deleteRemote(teamUid: String) {
        val uid = currentUserId
        try {
            SupabaseManager.client.postgrest
                .from("team_favorites")
                .delete {
                    filter {
                        eq("team_uid", teamUid)
                        if (uid != null) {
                            eq("user_id", uid)
                        } else {
                            eq("device_id", deviceId)
                            exact("user_id", null)
                        }
                    }
                }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            Log.e(TAG, "delete failed: ${e.message}", e)
        }
    }

    private fun loadFromDisk(): Map<String, FavoriteRow> {
        val raw = prefs.getString(KEY, null) ?: return emptyMap()
        return try {
            json.decodeFromString<List<FavoriteRow>>(raw).associateBy { it.teamUid }
        } catch (_: Exception) {
            emptyMap()
        }
    }

    private fun saveToDisk(map: Map<String, FavoriteRow>) {
        try {
            val encoded = json.encodeToString(map.values.toList())
            prefs.edit().putString(KEY, encoded).apply()
        } catch (_: Exception) {
            // Persisting favorites is best-effort; ignore serialization failures.
        }
    }

    /**
     * Clears all in-memory team-favorite state and removes the local disk
     * cache. Called from [AuthViewModel.signOut] so the next user starts with
     * an empty favorites set instead of inheriting the previous user's teams.
     * Deliberately does NOT delete server rows.
     */
    fun clearLocalCache() {
        _rows.value = emptyMap()
        prefs.edit().remove(KEY).apply()
    }

    companion object {
        private const val TAG = "TeamFavorites"
        private const val KEY = "gs.teamFavorites.rows"

        @Volatile private var instance: TeamFavoritesService? = null
        fun init(context: Context): TeamFavoritesService =
            instance ?: synchronized(this) {
                instance ?: TeamFavoritesService(context.applicationContext).also { instance = it }
            }
        fun get(): TeamFavoritesService =
            instance ?: error("TeamFavoritesService not initialized")
    }
}
